pvanalytics/features/snow.py

```diff
@@ -0,0 +1,266 @@
+        DC voltage modeled using measured plane-of-array irradiance and
+        back-of-module temperature. Assumes transmission equals 1. [V]
+    min_dcv : float
+        The lower voltage bound on the inverter's maximum power point
+        tracking (MPPT) algorithm. [V]
+    max_dcv : float
+        The upper voltage bound on the inverter's maximum power point
+        tracking (MPPT) algorithm. [V]
+    threshold_vratio : float
+        The lower bound for vratio that is representative of snow-free
+        conditions. Determined empirically. Depends on system configuration and
+        site conditions. [unitless]
+    threshold_transmission : float
+        The lower bound on transmission that is found under snow-free
+        conditions, determined empirically. [unitless]
+
+    Returns
+    -------
+    mode : int or None
+        ``mode`` is ``None`` when any of the inputs used to determine ``mode``
+        is ``nan``.
+
+    References
+    ----------
+    .. [1] E. C. Cooper, J. L. Braid and L. M. Burnham, "Identifying the
+       Electrical Signature of Snow in Photovoltaic Inverter Data," 2023 IEEE
+       50th Photovoltaic Specialists Conference (PVSC), San Juan, PR, USA,
+       2023, pp. 1-5, :doi:`10.1109/PVSC48320.2023.10360065`.
+    """
+    umin_meas = measured_voltage >= min_dcv
+    umax_meas = measured_voltage < max_dcv
+
+    umin_model = modeled_voltage_with_ideal_transmission >= min_dcv
+    umax_model = modeled_voltage_with_ideal_transmission < max_dcv
+
+    # Voltage is modeled as NaN if T = 0, but V = 0 makes more sense
+    modeled_voltage_with_calculated_transmission[transmission == 0] = 0
+
+    with np.errstate(divide='ignore'):
+        vmp_ratio =\
+            measured_voltage / modeled_voltage_with_calculated_transmission
+
+    # take care of divide by zero
+    vmp_ratio[modeled_voltage_with_calculated_transmission == 0] = 1
+
+    # vmp_ratio discrimates between states (1,2) and (3,4)
+    uvr = np.where(vmp_ratio >= threshold_vratio, 3, 1)
+
+    # transmission discrimates within (1,2) and (3,4)
+    utrans = np.where(transmission >= threshold_transmission, 1, 0)
+
+    # None if transmission, vmp_ratio, modeled_voltage_with_ideal_transmission,
+    # modeled_voltage_with_calculated_transmission, or measured_voltage is nan
+    # -1 if umin_model is 0
+    # 0 if umin_meas is 0, i.e., measurement indicate no power but it must be
+    # that umin_model is 1
+    # state 1, 2, 3, 4 defined by uvr + utrans
+    # -1 if umax_model is 0
+    # -1 if umax_meas is 0
+
+    mode = umin_meas * (uvr + utrans)
+    mode = np.where(~umin_model | ~umax_model | ~umax_meas, -1, mode)
+    mode = np.where(np.isnan(vmp_ratio) | np.isnan(transmission) |
+                    np.isnan(measured_voltage) |
+                    np.isnan(modeled_voltage_with_calculated_transmission) |
+                    np.isnan(modeled_voltage_with_ideal_transmission),
+                    None, mode)
+
+    return mode, vmp_ratio
```

Comment:
If this function needs to return `vmp_ratio` (does it?), the `Returns` section of the docstring should list it

Reply:
Added vmp_ratio to docstring

Reply:
Added` vmp_ratio` to the `Returns` section of the docstring